Add github dark

steff456 opened the pull request:
This PR is part of https://github.com/Quansight-Labs/jupyterlab-accessible-themes/issues/40

For adding a new theme a new directory is created with the following structure,

```
├── theme_name
│   ├── src
│   │   ├── index.ts
│   ├── style
│   │   ├── index.css
│   │   ├── variables.css
│   ├── package.json
│   ├── tsconfig.json
│   ├── README.md
```

A JupyterLab instance will start in your browser and you can select the theme, by clicking `Settings > Theme > Github Dark` in the main menu.

Some screenshots on how it looks right now,

<img width="1792" alt="image" src="https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/20992645/e4471e23-2137-42dd-b22f-4d1c19a5e082">

New theme checklist:

- [x]  New directory with new package
- [x] Update the setup.py to include the new package
- [x] Create a README.md inside the new package with information about the color palette and resources
- [x] Update main README.md with new theme

trallard commented:
@steff456 can you rebase against main? or merge main? there are merge conflicts here

gabalafou reviewed (changes requested):
Yay! I was able to install and load this theme! I like it!

My one concern is that to my eyes it looks like a lot of contrast has been "washed out" with this theme. Here's a screenshot of the launcher, for example:

<img width="662" alt="" src="https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/317883/7bdf6910-f2f2-4722-915b-8b7c4a56649f">

I can barely tell where the background ends and any given launcher card begins. And there are a few other places that appear low contrast to my eyes.

Is this something we need to fix now or later?

I'm marking this "request changes" because even if now is not the time to modify this GitHub Dark theme, the question about when or if it needs to be fixed needs to be resolved before I feel good about hitting the approve button. Thanks!

gabalafou commented on README.md:
```diff
@@ -41,6 +41,7 @@ This repository defines a set of accessible themes according to [WCAG color stan
 
 - [Pitaya Smoothie](./packages/pitayasmoothie/README.md) - Color-blind friendly
 - [GitHub light](./packages/githublight/README.md) - Color-blind friendly
```
```suggestion
- [GitHub Light](./packages/githublight/README.md) - Color-blind friendly
```

or `GitHub dark`, but we should be consistent, either way

trallard commented:
> Is this something we need to fix now or later?

This should be fixed now rather than later cc/ @steff456

steff456 commented:
I updated the color of all the borders, just for them to be more visible. Now the interface looks like this,

<img width="1792" alt="image" src="https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/20992645/25e68fea-252f-46d3-abca-e78c4e8a61b4">

trallard commented:
In the README and the conda-forge feedstock, we are requesting Python >= 3.6. See #56 3.7 has already reached EOL so we should be requesting 3.8 or above really

trallard commented:
Noticed again that the schema error pops with this theme too (see #57)

There are also several issues I have identified in #59 that need fixing

trallard commented on README.md:
```diff
@@ -40,7 +40,8 @@ This repository defines a set of accessible themes according to [WCAG color stan
 - [Pitaya Smoothie](./packages/pitayasmoothie/README.md) - Color-blind friendly
-- [GitHub light](./packages/githublight/README.md) - Color-blind friendly
+- [GitHub Light](./packages/githublight/README.md) - Color-blind friendly
+- [Github Dark](./packages/githubdark/README.md) - Color-blind friendly
```
Is this so? There are both GH dark and GH dark colourblind friendly https://github.com/primer/github-vscode-theme if the latter then it seems the name of the theme should be corrected

steff456 replied:
In the VSCode documentation it states that all the themes support different types of color blindness, it is not specific to which one though 🤷‍♀️

Do you prefer to only add the `color-blind friendly` note to the themes that are named specifically?

trallard replied:
Unless a theme is guaranteed to be colourblind-friendly, we should not say it is (otherwise, folks will expect this to be the case).

From the VSCode extension, https://github.com/primer/github-vscode-theme#install, some themes are colourblind friendly - if in this PR you *are not* implementing any of those marked as colourblind friendly, then this should not be called out. (my understanding was you were adding the light and dark default ones)

> GitHub Light Default
GitHub Light High Contrast ✨ new ✨
GitHub Light Colorblind ✨ new ✨
GitHub Dark Default
GitHub Dark High Contrast
GitHub Dark Colorblind ✨ new ✨
GitHub Dark Dimmed

steff456 replied:
ok! I'll make the change

steff456 commented:
> In the README and the conda-forge feedstock, we are requesting Python >= 3.6. See https://github.com/Quansight-Labs/jupyterlab-accessible-themes/issues/56 3.7 has already reached EOL so we should be requesting 3.8 or above really

I saw this item in the tracking issue for the next release and will make a separate PR fixing it.

steff456 commented:
@trallard the last commits fixed the issues for this theme of both https://github.com/Quansight-Labs/jupyterlab-accessible-themes/issues/59 and https://github.com/Quansight-Labs/jupyterlab-accessible-themes/issues/57. Please check it and let me know if anything else is needed to merge this PR 🙃

trallard commented:
Seems the schema error was not fixed here see:

<details>
```bash
[W 2023-07-07 14:10:04.166 LabApp] wrote error: 'Schema not found: /Users/tania/mambaforge/envs/jlab-themes/share/jupyter/lab/schemas/jupyterlab-accessible-themes/pitayasmoothie.json'
    Traceback (most recent call last):
      File "/Users/tania/mambaforge/envs/jlab-themes/lib/python3.10/site-packages/tornado/web.py", line 1784, in _execute
        result = method(*self.path_args, **self.path_kwargs)
      File "/Users/tania/mambaforge/envs/jlab-themes/lib/python3.10/site-packages/tornado/web.py", line 3290, in wrapper
        return method(self, *args, **kwargs)
      File "/Users/tania/mambaforge/envs/jlab-themes/lib/python3.10/site-packages/jupyterlab_server/settings_handler.py", line 48, in get
        result, warnings = get_settings(
      File "/Users/tania/mambaforge/envs/jlab-themes/lib/python3.10/site-packages/jupyterlab_server/settings_utils.py", line 370, in get_settings
        schema, version = _get_schema(schemas_dir, schema_name, overrides, labextensions_path)
      File "/Users/tania/mambaforge/envs/jlab-themes/lib/python3.10/site-packages/jupyterlab_server/settings_utils.py", line 47, in _get_schema
        raise web.HTTPError(404, notfound_error % path)
    tornado.web.HTTPError: HTTP 404: Not Found (Schema not found: /Users/tania/mambaforge/envs/jlab-themes/share/jupyter/lab/schemas/jupyterlab-accessible-themes/pitayasmoothie.json)
[W 2023-07-07 14:10:04.167 LabApp] 404 GET /lab/api/settings/jupyterlab-accessible-themes:pitayasmoothie?1688735404102 (e23722994c7b43e983d7df74432af902@::1) 17.01ms referer=http://localhost:8888/lab
```
</details>

but there is another PR named "Fix schema error" #61 - can you confirm which addresses the schema error in #57?

trallard commented:
Some issues that have yet to be fixed:

- [ ] Selected file colours are still not WCAG compliant - there should be a 3:1 contrast between `.jp-DirListing-item.jp-mod-selected` and the background and 4.5:1 between text and `.jp-DirListing-item.jp-mod-selected`

![Colour_Contrast_Analyser__CCA_](https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/23552331/131cbe2c-8b1a-422b-9126-a2305aece366)

- [ ] Same issue on hover on menu items hover on background fails to meet 3:1

![Colour_Contrast_Analyser__CCA__and_lab_-_JupyterLab](https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/23552331/688cd97d-f624-4857-868b-b9e07368c911)

- [ ] dividers are not noticeable

![lab_-_JupyterLab](https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/23552331/05842e03-8c43-4f7c-80b9-fc6a8f5ad5b6)

- [ ] Icons not having enough contrast against background

![lab_-_JupyterLab](https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/23552331/55a117f2-81d9-4b4a-bf1a-0aae85c294b5)

![Colour_Contrast_Analyser__CCA__and_JupyterLab](https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/23552331/81dacc3a-bb86-43b2-ac7a-3db397baf586)

- [ ] For whatever reason the dividers in the menus are hoverable when they are not in the default JLab themes, this seems to be a bug


![Untitled_ipynb_-_JupyterLab](https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/23552331/4a98ba1b-ea3a-4d1d-ba65-09ccfefb56da)
![Untitled_ipynb_-_JupyterLab](https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/23552331/ab823c51-db36-4607-93fa-9b2143a30517)


- [ ] New tab button does not have enough contrast against background 
![Colour_Contrast_Analyser__CCA__and_Untitled_ipynb_-_JupyterLab](https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/23552331/ce7785dd-77c8-4325-94dd-7c406a8496e5)

@steff456 I suggest you check the colours with a contrast checker or a web extension to ensure the colours meet WCAG AA standards

steff456 commented:
@trallard The last commit addresses all the issues you found. 

I tried the suggestion @gabalafou gave me last week of trying lighthouse and axe chrome extensions to test for colors, but the tools didn't flag any color contrast issues. 

Please play around with the theme and let me know if there's something else that's off. I'm testing as much as I can on my end but there's a lot of parts that I may be missing.

gabalafou reviewed (changes requested):
@steff456 I see what you mean about Axe Chrome Dev Tools not catching contrast issues.

Here's a screenshot of the toast notification asking the user if they would like to receive JupyterLab news. With this GitHub Dark theme applied, the background color of the buttons (open privacy policy, yes, no) is indistinguishable from the toast background, so they don't even look like buttons: 

<img width="364" alt="" src="https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/317883/624db6e6-4253-4960-8f53-b8e1b063f2c3">

But Axe didn't catch that. 

It also didn't catch what I bet is a problematic contrast issue between the yaml file icon and the file browser background:

<img width="291" alt="filebrowser screenshot showing poorly contrasting yaml file" src="https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/317883/149b550b-4730-4bb3-8700-2aa41498b363">

Nor did it catch the lack of contrast between the menubar color and the active menu item color. 

Another thing—not contrast-related—is that some of the vertical and horizontal divider lines in the UI have been messed up. In following screenshot, for example, if you look closely in the middle you can see a line that sticks out 1 to 2 pixels too far:

<img width="126" alt="" src="https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/317883/85dd78d2-6d73-493e-9616-cb0b255fd4cf">

In the other themes, I think the border color matches the background in certain spots, so you don't see these little visual bugs.

I also can't seem to open the property inspector or the debugger from the right side bar. Are you seeing the same?

gabalafou commented:
Figured out the issue with the property inspector and the debugger. With GitHub Dark theme, they don't have minimum widths, so I can drag them to zero width.

steff456 commented:
The latest changes fix the issues you flagged @gabalafou 

- [x] Notification center buttons are now visible

<img width="401" alt="image" src="https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/20992645/d902102d-af0c-4471-a230-b02ca82f6d3b">

![image](https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/20992645/711502fd-f445-4953-a31a-4b9231185eed)

- [x] Contrast issue between some icons in the files menu and the background

<img width="667" alt="image" src="https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/20992645/b1b2e681-1b6c-4fa2-8186-54d1131b285d">

- [x] Add a minimum sidebar width value

gabalafou commented:
Not sure if the following problems should block merging this PR but I'm noting them down so they're recorded somewhere.

- [ ] (1) Not enough contrast between the blue `button:focus-visible` focus ring and the dark background: <img width="47" alt="image" src="https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/317883/06950b26-e8c1-409e-957f-43d70b324508">

<img width="592" alt="Color Contrast Analyzer app shows that outline blue of #0D47A1 against black #212121 has 1.9:1 contrast ratio which fails every WCAG 2.1 contrast requirement" src="https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/317883/14e32957-58eb-4ae5-95f8-cc631dbafdc6">

Side note: JupyterLab Dark has the same low contrast problem with the focus ring.

- [ ] (2) No visible focus indicator for buttons in toast notification:

<img width="483" alt="" src="https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/317883/3254ef74-4121-4970-a3e3-eee122edb994">

On inspection the issue seems to be the following CSS rule:

```css
.jp-toast-button:focus {
    outline: 1px solid var(--jp-reject-color-normal, var(--jp-layout-color2));
}
```

Dev Tools says that `--jp-reject-color-normal` is not defined, whereas the fallback `--jp-layout-color2` is set to `--gh-dark-gray8: #161b22` which is the same color that toast background is set to: 

```css
background: var(--toastify-color-light);
--toastify-color-light: var(--jp-layout-color1);
--jp-layout-color1: var(--gh-dark-gray8);
--gh-dark-gray8: #161b22;
```

Not sure if the issue here is that --jp-layout-color2 and --jp-layout-color-1 are the same, or if the issue is `--jp-reject-color-normal` not being defined, or the `Toastify__toast-theme--light` class getting applied when this is a dark theme, or something else.

- [ ] (3) When hovered, the "Do Not Check For Updates" button in the toast notification seems to have a transparent background, so it ends up looking like text instead of a button when hovered:

![](https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/317883/252ae979-dfc0-4eb8-8b2c-63299e4c6f53)

- [ ] (4) Not enough contrast for green status dot for notebook file in file browser: ![](https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/317883/912e253f-e943-4727-9740-c0c486702c71)

<img width="592" alt="Color Contrast App showing that green #196C2E against black #161B22 has contrast ratio 2.7:1 and meets no WCAG 2.1 contrast requirements" src="https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/317883/4f643a57-d580-4a9b-b0a3-33b5db768310">

- [ ] (5) Not enough contrast for package names in License viewer:  ![localhost_8888_lab_workspaces_auto-V_tree_jupyterlab_accessible_themes](https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/317883/3cba4579-34c5-4d23-85a8-d96dd4ba2185)

<img width="592" alt="image" src="https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/317883/09b1daa3-d46d-4da2-a918-8d0d7f593ab5">

I'm going to stop there. There are probably other things I didn't catch, but overall, at time of this review, the theme seems in pretty good shape to me.

trallard commented:
ok to get this PR in, I would say let's fix the following (they seem scoped and small enough)

- [ ] (4) Not enough contrast for green status dot for notebook file in the file browser
- [ ] (3) When hovered, the "Do Not Check For Updates" button in the toast notification seems to have a transparent background, so it ends up looking like text instead of a button when hovered

The other items: @steff456  can you move these to a new issue so we can keep track just in case?
Once those are sorted we will merge

steff456 commented:
The last commit fixes all the issues raised by @gabalafou,

1. Now the focus indicator has the same blue as in other places,

![image](https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/20992645/ac46dc24-2453-4adf-a5f1-4a3c69e7b313)

![image](https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/20992645/62e8e913-d888-46c1-8591-646bdb61a469)

2. Visible focus indicator for buttons in toast notifications, (I left the same combination as above)

![image](https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/20992645/e0a8b574-6acc-4fcb-b463-1c9ad4b7f462)

3. Fixed the hover state of the buttons in the toast notifications, now it will use a lighter shade of gray

![image](https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/20992645/d4832ecd-f5e2-4ec6-887f-31bbfce854ba)

4. Fixed the contrast for the green status for the notebook icon in the file browser,

![image](https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/20992645/256b280b-4b76-4b9c-ba96-2324385ec519)

![image](https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/20992645/066d261c-1bfe-4b6f-8ec1-c54035493145)

5. Fixed the contrast for package names in License viewer

![image](https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/20992645/87ab91ab-d7bd-4e8b-bce8-5038b2573213)

![image](https://github.com/Quansight-Labs/jupyterlab-accessible-themes/assets/20992645/bd3abcb1-38cc-4877-8b53-3a0c06573474)

trallard commented:
Thanks @steff456 this looks good to me. 
Will wait for @gabalafou to confirm but I think this is in a good shape to merge now.

gabalafou reviewed (approved):
Sorry I didn't realize this was waiting on me to push the button! :D

trallard commented:
Thanks folks. @steff456 this is ready to merge now, thanks for seeing this lengthy contribution through 🙏🏽

steff456 commented:
Thanks!! I'm super happy to see this finally merged
